### ISSUE TITLE
(Node.js) Error: Cannot read property '0' of null

### ISSUE THREAD

**Issue by bergkvist** [tirithen/node-jsoncompress#2]
The following code causes an error to be thrown:

```javascript
import jsoncompress from 'jsoncompress'

const compressionTemplate = {
    id: 0,
    signature: '', 
    orders: { 
        pending: [{ floor: 0, type: 0 }], 
        accepted: [{ floor: 0, type: 0 }], 
        completed: [{ floor: 0, type: 0 }] 
    },
    currOrder: { floor: 0, type: 0 }
}

const data = {
    id: 0,
    signature: 'TTM',
    orders: {
        pending: [],
        accepted: [{ floor: 2, type: 1 }],
        completed: []
    },
    currOrder: { floor: 2, type: 1 }
}

const compressedData = jsoncompress.compress(data, compressionTemplate)
const buffer = new Buffer(compressedData)
const str = buffer.toString()
const decompressed = jsoncompress.decompress(str, compressionTemplate) 
// ^ This throws with 'Cannot read property '0' of null'
```

However, changing the data very slightly will make it work:
```javascript
import jsoncompress from 'jsoncompress'

const compressionTemplate = {
    id: 0,
    signature: '', 
    orders: { 
        pending: [{ floor: 0, type: 0 }], 
        accepted: [{ floor: 0, type: 0 }], 
        completed: [{ floor: 0, type: 0 }] 
    },
    currOrder: { floor: 0, type: 0 }
}

const data = {
    id: 3,  // <------------ This is the only change!
    signature: 'TTM',
    orders: {
        pending: [],
        accepted: [{ floor: 2, type: 1 }],
        completed: []
    },
    currOrder: { floor: 2, type: 1 }
}

const compressedData = jsoncompress.compress(data, compressionTemplate)
const buffer = new Buffer(compressedData)
const str = buffer.toString()
const decompressed = jsoncompress.decompress(str, compressionTemplate) 
// ^ Correctly returns the data
```

NOTE: omitting the buffer will make this work, so I'm assuming this might have something to do with the string encoding?

**Comment by bergkvist** [comment]
UPDATE: I discovered what I believe is the problem.

https://stackoverflow.com/questions/17006799/how-do-i-capture-utf-8-decode-errors-in-node-js

One of the characters in the string (resulting from the compression) is not a valid utf8-character, hence it is replaced by the "REPLACEMENT CHARACTER" (U+FFFD), when converting the string to a buffer.

**Comment by bergkvist** [comment]
### So it turns out that a few things went wrong here:
1. JavaScript strings are encoded with utf-16 (unicode)
2. `Buffer.from(string)` only accepts utf-8 strings by default. 
3. `Buffer.from(string)` does not throw an error when it receives something invalid, but rather changes it into something that is valid. (This might have been "fixed" in newer versions of node)
4. The output of `jsoncompress.compress(data, template)` is a string utilizing the entire utf-16 space. This means certain resulting characters might not be valid utf-8 characters.

---

When the buffer object receives an invalid character, it will replace it with the *REPLACEMENT CHARACTER* (`\ufffd`), and it will do so silently.

Notice that utf-8 and utf-16 are equal "most of the time", which means certain people might never experience this problem, based on the diversity of messages they are sending.

---

### The solution
Make sure to tell the Buffer object that you are using a different encoding than "default". In this case, utf-16:
```javascript
const utf16String = jsoncompress.compress(data, template)
const utf16Buffer = Buffer.from(utf16String, 'utf16le')  // <-- Notice the second argument
```
You also need to specify the encoding when you convert the buffer back to a string.
```javascript
const utf16String = utf16Buffer.toString('utf16le')
const data = jsoncompress.decompress(utf16String, template)
```

In other words, specifying `'utf16le'` as encoding when creating/stringifying the buffers will solve the problem.
